pkg/controller/session/session_controller_test.go

```diff
@@ -0,0 +1,321 @@
+			BeforeEach(func() {
+				objects = []runtime.Object{
+					&v1alpha1.Session{
+						ObjectMeta: metav1.ObjectMeta{
+							Name:      "test-session",
+							Namespace: "test",
+						},
+						Spec: v1alpha1.SessionSpec{
+							Refs: []string{"details"},
+						},
+					},
+				}
+			})
+
+			It("finalizer added", func() {
+				res, err := controller.Reconcile(req)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(res.Requeue).To(BeFalse())
+
+				modified := GetSession("test", "test-session")
+				Expect(modified.ObjectMeta.Finalizers).To(HaveLen(1))
+			})
+
+			It("found locator", func() {
+				locator.Action = foundTestLocator
+
+				res, err := controller.Reconcile(req)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(res.Requeue).To(BeFalse())
+
+				Expect(locator.WasCalled).To(BeTrue())
+				Expect(mutator.WasCalled).To(BeTrue())
+			})
+
+			It("status created", func() {
+				locator.Action = foundTestLocator
+				mutator.Action = basicTestMutator(model.ResourceStatus{Name: "details", Kind: "test", Action: model.ActionCreated})
+
+				res, err := controller.Reconcile(req)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(res.Requeue).To(BeFalse())
+
+				Expect(locator.WasCalled).To(BeTrue())
+				Expect(mutator.WasCalled).To(BeTrue())
+				Expect(revertor.WasCalled).ToNot(BeTrue())
+
+				modified := GetSession("test", "test-session")
+				Expect(modified.Status).ToNot(BeNil())
+				Expect(modified.Status.Refs).To(HaveLen(1))
+				Expect(modified.Status.Refs[0].Name).To(Equal("details"))
+			})
+		})
+	})
+
+	Context("session modification", func() {
+		Context("new reference", func() {
+			BeforeEach(func() {
+				objects = []runtime.Object{
+					&v1alpha1.Session{
+						ObjectMeta: metav1.ObjectMeta{
+							Name:       "test-session",
+							Namespace:  "test",
+							Finalizers: []string{session.Finalizer},
+						},
+						Spec: v1alpha1.SessionSpec{
+							Refs: []string{"details", "details2"},
+						},
+						Status: v1alpha1.SessionStatus{
+							Refs: []*v1alpha1.RefStatus{
+								{
+									Name:      "details",
+									Resources: []*v1alpha1.RefResource{{Kind: &kind, Name: &name, Action: &action}},
+								},
+							},
+						},
+					},
+				}
+			})
+			It("status updated", func() {
+				locator.Action = foundTestLocator
+				mutator.Action = basicTestMutator(model.ResourceStatus{Name: "details2", Kind: "test", Action: model.ActionCreated})
+
+				res, err := controller.Reconcile(req)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(res.Requeue).To(BeFalse())
+
+				Expect(locator.WasCalled).To(BeTrue())
+				Expect(mutator.WasCalled).To(BeTrue())
+				Expect(revertor.WasCalled).ToNot(BeTrue())
+
+				modified := GetSession("test", "test-session")
+				Expect(modified.Status).ToNot(BeNil())
+				Expect(modified.Status.Refs).To(HaveLen(2))
+				Expect(modified.Status.Refs[0].Name).To(Equal("details"))
+				Expect(modified.Status.Refs[1].Name).To(Equal("details2"))
```

Comment:
Maybe we can transform refs extracting names and assert on slice instead? Not sure that improves readability much though... just a thought.

Reply:
Was wondering about it... but...

Reply:
Yeah, let's leave that